File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int[] dp = new int[N + 1];
		int[] stairs = new int[N + 1];

		for (int i = 1; i <= N; i++) {
			stairs[i] = Integer.parseInt(br.readLine());
		}

		dp[1] = stairs[1];

		if (N > 1) {
			dp[2] = stairs[1] + stairs[2];
		}

		for (int i = 3; i <= N; i++) {
			dp[i] = Math.max(dp[i - 2] + stairs[i], dp[i - 3] + stairs[i - 1] + stairs[i]);
		}

		System.out.println(dp[N]);
		br.close();
	}
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.lang.Integer.max

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.out))

fun main() {
    val n = br.readLine().toInt()
    val dp = Array(n) { 0 }
    val arr = IntArray(n) { br.readLine().toInt() }

    dp[0] = arr[0]
    if (n >= 2) {
        dp[1] = arr[0] + arr[1]
        if (n >= 3) {
            dp[2] = max(arr[0] + arr[2], arr[1] + arr[2])

            for (i in 3 until n) {
                dp[i] = max(dp[i - 2] + arr[i], dp[i - 3] + arr[i - 1] + arr[i])
            }
        }
    }

    bw.write("${dp[n - 1]}")
    bw.flush()
    bw.close()
    br.close()
}